### ISSUE TITLE
Rsyslog delays shutdown when it can't connect to remote logger

### ISSUE THREAD

**Issue by gaia** [rsyslog/rsyslog#3812]
I have rsyslog in the hypervisor logging remotely. The hypervisor's conectivity is managed by one of it's guests, which naturally gets shutdown by the hypervisor before the hypervisor starts the reboot routines. Therefore when rebooting the hypervisor, rsyslog loses connectivity and I get a long delay on shutting down rsyslog, presumably because it is trying to send data before exiting.

Is there a setting to tell rsyslog to give up quicker when there is no connection? or better yet, to exit immediately and save to disk what it needs to send, so that it can send when the connection becomes available again?

This is Debian 10 with rsyslogd  8.1901.0

**Comment by davidelang** [comment]
can you instrument your system to see if it's trying to do DNS lookups during 
the shutdown? I'm suspecting that's more of a problem than sending.

there is the saveonshutdown option if you have a disk assisted queue

if you can get a debuglog during shutdown it would tell us exactly what is 
happening.

a config would also help

David Lang

**Comment by gaia** [comment]
Thanks David.

The config is below. There is nothing in the log files. I will look into into instrumenting the system, but regardless of whether it is data or DNS lookups, how do I enable a global saveonshutdown ?

```
# /etc/rsyslog.conf configuration file for rsyslog
#
# For more information install rsyslog-doc and see
# /usr/share/doc/rsyslog-doc/html/configuration/index.html


#################
#### MODULES ####
#################

module(load="imuxsock") # provides support for local system logging
module(load="imklog")   # provides kernel logging support
#module(load="immark")  # provides --MARK-- message capability

# provides UDP syslog reception
#module(load="imudp")
#input(type="imudp" port="514")

# provides TCP syslog reception
#module(load="imtcp")
#input(type="imtcp" port="514")


###########################
#### GLOBAL DIRECTIVES ####
###########################

#
# Use traditional timestamp format.
# To enable high precision timestamps, comment out the following line.
#
$ActionFileDefaultTemplate RSYSLOG_TraditionalFileFormat

#
# Set the default permissions for all log files.
#
$FileOwner root
$FileGroup adm
$FileCreateMode 0640
$DirCreateMode 0755
$Umask 0022

#
# Where to place spool and state files
#
$WorkDirectory /var/spool/rsyslog

#
# Include all config files in /etc/rsyslog.d/
#
$IncludeConfig /etc/rsyslog.d/*.conf


###############
#### RULES ####
###############

#
# First some standard log files.  Log by facility.
#
auth,authpriv.*			/var/log/auth.log
*.*;auth,authpriv.none		-/var/log/syslog
#cron.*				/var/log/cron.log
daemon.*			-/var/log/daemon.log
kern.*				-/var/log/kern.log
lpr.*				-/var/log/lpr.log
#mail.*				-/var/log/mail.log
user.*				-/var/log/user.log

#
# Logging for the mail system.  Split it up so that
# it is easy to write scripts to parse these files.
#
mail.info			-/var/log/mail.info
mail.warn			-/var/log/mail.warn
mail.err			/var/log/mail.err

#
# Some "catch-all" log files.
#
*.=debug;\
	auth,authpriv.none;\
	news.none;mail.none	-/var/log/debug
*.=info;*.=notice;*.=warn;\
	auth,authpriv.none;\
	cron,daemon.none;\
	mail,news.none		-/var/log/messages

#
# Emergencies are sent to everybody logged in.
#
*.emerg				:omusrmsg:*

# Papertrail
$DefaultNetstreamDriverCAFile /etc/papertrail-bundle.pem # trust these CAs
$ActionSendStreamDriver gtls # use gtls netstream driver
$ActionSendStreamDriverMode 1 # require TLS
$ActionSendStreamDriverAuthMode x509/name # authenticate by hostname
$ActionSendStreamDriverPermittedPeer *.papertrailapp.com

*.*    @@logs.papertrailapp.com:38549
```

**Comment by davidelang** [comment]
nothing in your config is doing anything to save messages when you shutdown, it 
will just throw away the contents of the main queue, so that's not what's 
causeing the slow shutdown.

I don't see anything obviously odd with this config.

**Comment by gaia** [comment]
Thanks @davidelang, I am going to try to figure out how to instrument the system to find out what is the cause of the delay on shutdown.

**Comment by davidelang** [comment]
you need to set saveonshutdown on each queue individually, if you want to change 
settings on the main queue, use main_queue()

David Lang

**Comment by gaia** [comment]
i got a queue setup

```
$ActionResumeInterval 10
$ActionQueueSize 100000
$ActionQueueDiscardMark 97500
$ActionQueueHighWaterMark 80000
$ActionQueueType LinkedList
$ActionQueueFileName fwdqueue
$ActionQueueCheckpointInterval 100
$ActionQueueMaxDiskSpace 2g
$ActionResumeRetryCount -1
$ActionQueueSaveOnShutdown on
$ActionQueueTimeoutEnqueue 2
$ActionQueueDiscardSeverity 0
```

I take it is only used when it can't fwd to the remote, cause fwdqueue isn't in the work dir.

**Comment by davidelang** [comment]
On Fri, 23 Aug 2019, gaia wrote:

> i got a queue setup
>
> ```
> $ActionResumeInterval 10
> $ActionQueueSize 100000
> $ActionQueueDiscardMark 97500
> $ActionQueueHighWaterMark 80000
> $ActionQueueType LinkedList
> $ActionQueueFileName fwdqueue
> $ActionQueueCheckpointInterval 100
> $ActionQueueMaxDiskSpace 2g
> $ActionResumeRetryCount -1
> $ActionQueueSaveOnShutdown on
> $ActionQueueTimeoutEnqueue 2
> $ActionQueueDiscardSeverity 0
> ```
>
> I take it is only used when it can't fwd to the remote, cause fwdqueue isn't in the work dir.

right, it only uses the disk assist queue when the memory queue fills up, and 
that only happens when it can't deliver messages to the remote fast enough (and 
you have it configured to throw away some messages when it gets too full, so 
that would further delay when anything would end up getting written to the disk 
queue)

We strongly recommend that you migrate to the new action() syntax rather than 
the multi-line format that you show here. We don't intend to break it any time 
soon (if ever), but it is much clearer what is happening with the new syntax

David Lang

**Comment by gaia** [comment]
I turned on logging during shutdown and this is what I found

```
# journalctl | grep "Sep 03 12:" | egrep -i "system logging service|rsyslog"
Sep 03 12:16:25 node2 systemd[1]: Stopping System Logging Service...
Sep 03 12:16:25 node2 rsyslogd[1766]:  [origin software="rsyslogd" swVersion="8.1901.0" x-pid="1766" x-info="https://www.rsyslog.com"] exiting on signal 15.
Sep 03 12:17:55 node2 systemd[1]: rsyslog.service: State 'stop-sigterm' timed out. Killing.
Sep 03 12:17:55 node2 systemd[1]: rsyslog.service: Killing process 1766 (rsyslogd) with signal SIGKILL.
Sep 03 12:17:55 node2 systemd[1]: rsyslog.service: Main process exited, code=killed, status=9/KILL
Sep 03 12:17:55 node2 systemd[1]: rsyslog.service: Failed with result 'timeout'.
Sep 03 12:17:55 node2 systemd[1]: Stopped System Logging Service.
```

**Comment by davidelang** [comment]
can you enable debugging during shutdown and send us that report?

David Lang

**Comment by gaia** [comment]
sorry, how would I do that?

**Comment by davidelang** [comment]
sorry, don't have time to go into details right now, which is why I'm just 
throwing a couple of links at you

https://www.rsyslog.com/doc/v8-stable/troubleshooting/debug.html
https://www.rsyslog.com/how-to-use-debug-on-demand/

you probably want to setup debug on demand and then enable it just before doing 
the shutdown

David Lang

**Comment by gaia** [comment]
I followed the instructions, and logging begins, I can tail the file and see it working. But after the reboot the log is empty again.

**Comment by ahem** [comment]
We are seeing this issue also, when using basic  (using rsyslog v8.2006.0). Basically, if we disconnect the network and then shutdown rsyslog, then we see the rsyslog process hanging for 90 seconds before exiting. If we disconnect the network, then wait a couple of minutes and then exit we do not see the delay.

```
action(type="omfwd"
            protocol="tcp"
            target="xxxxx.xxx"
            port="6514"
            StreamDriver="gtls"
            StreamDriverMode="1"
            StreamDriverAuthMode="x509/name"
            StreamDriverPermittedPeers="*.xxxxx.xxx")
```

**Comment by rgerhards** [comment]
@ahem which version? If not current (8.2010.0) I suggest to update.

**Comment by rgerhards** [comment]
This looks like omfwd does not terminate, and rsyslog needs to resort to thread cancellation. The latter is always problematic, but IIRC there was a bug in the cancellation handler. Maybe that bug bites here. In any case, can you please provide your full config.

**Comment by rgerhards** [comment]
Just for clarity: it is that part showcasing the issue in worker termination:
```
1403.948153880:main thread    : action-4-builtin:omfwd queue: ../../rsyslog-8.2206.0/runtime/queue.c: regular queue workers shut down.
1403.948194920:main thread    : action-4-builtin:omfwd queue: ../../rsyslog-8.2206.0/runtime/queue.c: we have a DA queue (0x4fc478), requesting its shutdown.
1403.952451400:main thread    : action-4-builtin:omfwd queue: ../../rsyslog-8.2206.0/runtime/queue.c: trying shutdown of regular worker of DA queue
1403.953708800:action-4-builtin:omfwd queue[DA]:Reg/w0: ../../rsyslog-8.2206.0/runtime/../action.c: actionDoRetry: action-4-builtin:omfwd enter loop, iRetries=1, ResumeInRow 1
1403.953835480:action-4-builtin:omfwd queue[DA]:Reg/w0: ../../rsyslog-8.2206.0/tools/omfwd.c: omfwd: tryResume: pWrkrData 0xb5001608
1403.953911240:action-4-builtin:omfwd queue[DA]:Reg/w0: ../../rsyslog-8.2206.0/tools/omfwd.c: TCPSendInit CREATE
1403.955067600:main thread    : ../../rsyslog-8.2206.0/runtime/wti.c: sent SIGTTIN to worker thread 0xb634e2e0
1403.955944960:main thread    : ../../rsyslog-8.2206.0/runtime/wtp.c: action-4-builtin:omfwd queue[DA]:Reg: waiting 2ms on worker thread termination, 1 still running
1403.958352800:main thread    : ../../rsyslog-8.2206.0/runtime/wtp.c: action-4-builtin:omfwd queue[DA]:Reg: timeout waiting on worker thread termination
1403.959435040:main thread    : ../../rsyslog-8.2206.0/runtime/wti.c: sent SIGTTIN to worker thread 0xb634e2e0
1403.960102640:main thread    : ../../rsyslog-8.2206.0/runtime/errmsg.c: Called LogMsg, msg: action-4-builtin:omfwd queue: regular queue shutdown timed out on DA queue (this is OK, timeout was 10)
1403.960957120:main thread    : ../../rsyslog-8.2206.0/runtime/operatingstate.c: osf: MSG action-4-builtin:omfwd queue: regular queue shutdown timed out on DA queue (this is OK, timeout was 10): checking to see if we need to cancel any worker threads of the primary queue
1403.961639600:main thread    : action-4-builtin:omfwd queue: ../../rsyslog-8.2206.0/runtime/queue.c: checking to see if we need to cancel any worker threads of the DA queue
1403.961891720:main thread    : ../../rsyslog-8.2206.0/runtime/errmsg.c: Called LogMsg, msg: action-4-builtin:omfwd queue: need to do cooperative cancellation - some data may be lost, increase timeout?
1403.962129480:main thread    : ../../rsyslog-8.2206.0/runtime/operatingstate.c: osf: MSG action-4-builtin:omfwd queue: need to do cooperative cancellation - some data may be lost, increase timeout?: sent SIGTTIN to worker thread 0xb634e2e0, giving it a chance to terminate
1404.013871680:main thread    : ../../rsyslog-8.2206.0/runtime/errmsg.c: Called LogMsg, msg: action-4-builtin:omfwd queue: need to do hard cancellation
1404.014453680:main thread    : ../../rsyslog-8.2206.0/runtime/operatingstate.c: osf: MSG action-4-builtin:omfwd queue: need to do hard cancellation: cooperative worker termination failed, using cancellation...
1404.015363200:main thread    : wti 0x5057e8: ../../rsyslog-8.2206.0/runtime/wti.c: canceling worker thread
1404.015631160:main thread    : wti 0x5057e8: ../../rsyslog-8.2206.0/runtime/wti.c: waiting on termination, state 3
```

**Comment by joaofl** [comment]
> Just for clarity: it is that part showcasing the issue in worker termination:
> 

It is... The next thing to happen is the debug-log file getting truncated (which for me also looks like a bug/limitation).

Here is my config:
[rsyslog.conf.txt](https://github.com/rsyslog/rsyslog/files/11673524/rsyslog.conf.txt)

**Comment by rgerhards** [comment]
> The next thing to happen is the debug-log file getting truncated (which for me also looks like a bug/limitation).

Maybe it is kept inside local rsyslog buffers, but even if so (I doubt but have not checked), it would be very few messages. Maybe journal just removes part of it (also not that likely). IOW: we never had any problems with truncated debug logs before in the past 20 yrs ;-)

**Comment by joaofl** [comment]
> Maybe it is kept inside local rsyslog buffers, but even if so (I doubt but have not checked), it would be very few messages. Maybe journal just removes part of it (also not that likely). IOW: we never had any problems with truncated debug logs before in the past 20 yrs ;-)

I felt this was the expected behavior, to just remove it upon shutdown, which is what I am observing here. If not, please ignore my comment.

**Comment by rgerhards** [comment]
> I felt this was the expected behavior, to just remove the debug logs upon service stop, which is what I am observing here.

Actually not. In almost all cases you need the debug log after the run is completed. So rsyslog does not touch it. The log, however, is overwritten when the instance starts anew.

**Comment by rgerhards** [comment]
You have `queue.saveOnShutdown="on"` - just for a test, change that to `off`. This is often the case for cancellation by systemd, but there is no indication inside the log file that rsyslog is still busy persisting to disk.

Also, it would be good if you could enable impstats, with a very short reporting interval, so that we can see if any queues remain undrained.

**Comment by joaofl** [comment]
> You have `queue.saveOnShutdown="on"` - just for a test, change that to `off`. 

I tried that, and the behavior remained the same.

> Also, it would be good if you could enable impstats, with a very short reporting interval, so that we can see if any queues remain undrained.

I will check on that.


One test i did to check if the queue thread would ever be killed by the main thread, was to extend the systemd's timeout to very long, and yet, rsyslog never timed-out neither exited after 30+ minutes.

In the meanwhile, from the debug logs, it seems that there is a queue which is still working, and queuing logs. Can it be that this queue never gets the order to exit?

I also understood the reason why my logs get truncated. It is because when I stop rsyslog service via systemd, it will still be restarted by the `syslog.socket`, leading to the overwrite.

**Comment by joaofl** [comment]
>Also, it would be good if you could enable impstats, with a very short reporting interval, so that we can see if any queues remain undrained.

I enables the stats print every second, cleaned all the previous logs, and run `systemctl stop syslog.socket rsyslog.service`.

Find below the stats and debug logs on the corresponding time window. Apparently there is a queue that is never drained. 

[rsyslog.debug.txt](https://github.com/rsyslog/rsyslog/files/11723105/rsyslog.debug.txt)
[rsyslog.stats.txt](https://github.com/rsyslog/rsyslog/files/11723106/rsyslog.stats.txt)

**Comment by joaofl** [comment]
I've been a few days on it, and I am really having a hard time figuring anything out, apart from the fact that it really looks like a race condition/deadlock. But I cant figure out why yet.

I am also considering using some work-around.... Is there any way to force the sync to the disk on the DA queues, that would allow me to safely hard-kill the rsyslog?

@rgerhards , any tips?

**Comment by joaofl** [comment]
Unfortunately I could not determine the root cause, but I will try to conclude my findings here.

For the DA workers, they fall into a deadlock in the following loop:
![image](https://github.com/rsyslog/rsyslog/assets/8092452/77136319-9712-4c30-bbdc-94d1f772b2ab)

The DA threads do get signaled by the main thread to shutdown, and the signal is propagated just fine. But to be shutdown, another iteration in the loop is required to reach again the test in line 421, which never happens for the DA workers. I initially thought it could be stuck trying to establish connection with the remote server, but no. I commented that code and the problem persisted.

I just could not figure out what the DA threads are doing, except that they are "working". The issue is that they don't re-iterate this loop as the other workers. It is not pooling, but more like busy-waiting while working.

When I attach gdb to the process while looping, I can see the main and two "ghost" threads, without any information available about (only some disassembled code). 

I did notice though that when `action.resumeRetryCount="0"` then the DA workers go into IDLE, and if you kill rsyslog during while they are idling, the you can reach a graceful shutdown.
